stdlib/public/core/KeyPath.swift

```diff
@@ -0,0 +1,1739 @@
+      body = popRaw(size)
+    } else {
+      body = UnsafeRawBufferPointer(start: nil, count: 0)
+    }
+    let component = RawKeyPathComponent(header: header, body: body)
+    
+    // fetch type, which is in the buffer unless it's the final component
+    let nextType: Any.Type?
+    if data.count == 0 {
+      nextType = nil
+    } else {
+      if MemoryLayout<Any.Type>.size == 8 {
+        // Words in the key path buffer are 32-bit aligned
+        nextType = unsafeBitCast(pop((Int32, Int32).self),
+                                 to: Any.Type.self)
+      } else if MemoryLayout<Any.Type>.size == 4 {
+        nextType = pop(Any.Type.self)
+      } else {
+        _sanityCheckFailure("unexpected word size")
+      }
+    }
+    return (component, nextType)
+  }
+  
+  mutating func pop<T>(_ type: T.Type) -> T {
+    _sanityCheck(_isPOD(T.self), "should be POD")
+    let raw = popRaw(MemoryLayout<T>.size)
+    let resultBuf = UnsafeMutablePointer<T>.allocate(capacity: 1)
+    _memcpy(dest: resultBuf,
+            src: UnsafeMutableRawPointer(mutating: raw.baseAddress.unsafelyUnwrapped),
+            size: UInt(MemoryLayout<T>.size))
+    let result = resultBuf.pointee
+    resultBuf.deallocate(capacity: 1)
+    return result
+  }
+  mutating func popRaw(_ size: Int) -> UnsafeRawBufferPointer {
+    _sanityCheck(data.count >= size,
+                 "not enough space for next component?")
+    let result = UnsafeRawBufferPointer(start: data.baseAddress, count: size)
+    data = UnsafeRawBufferPointer(
+      start: data.baseAddress.unsafelyUnwrapped + size,
+      count: data.count - size
+    )
+    return result
+  }
+}
+
+// MARK: Library intrinsics for projecting key paths.
+
+public // COMPILER_INTRINSIC
+func _projectKeyPathReadOnly<Root, Value>(
+  root: Root,
+  keyPath: KeyPath<Root, Value>
+) -> Value {
+  return keyPath.projectReadOnly(from: root)
+}
+
+public // COMPILER_INTRINSIC
+func _projectKeyPathWritable<Root, Value>(
+  root: UnsafeMutablePointer<Root>,
+  keyPath: WritableKeyPath<Root, Value>
+) -> (UnsafeMutablePointer<Value>, Builtin.NativeObject) {
+  return keyPath.projectMutableAddress(from: root)
+}
+
+public // COMPILER_INTRINSIC
+func _projectKeyPathReferenceWritable<Root, Value>(
+  root: Root,
+  keyPath: ReferenceWritableKeyPath<Root, Value>
+) -> (UnsafeMutablePointer<Value>, Builtin.NativeObject) {
+  return keyPath.projectMutableAddress(from: root)
+}
+
+// MARK: Appending type system
+
+// FIXME(ABI): The type relationships between KeyPath append operands are tricky
+// and don't interact well with our overriding rules. Hack things by injecting
+// a bunch of `appending` overloads as protocol extensions so they aren't
+// constrained by being overrides, and so that we can use exact-type constraints
+// on `Self` to prevent dynamically-typed methods from being inherited by
+// statically-typed key paths.
+public protocol _AppendKeyPath {}
+
+extension _AppendKeyPath where Self == AnyKeyPath {
```

Comment:
What exactly doesn't work? It seems to do what I'm abusing it for, making the methods only available exactly on AnyKeyPath.